virttest/utils_test/__init__.py

```diff
@@ -1990,8 +1995,21 @@ def download_stress(self):
         url = self.download_url
         tmp_path = data_dir.get_tmp_dir()
         logging.info('Download stress tool from %s', url)
-        download_method = getattr(self, "_%s_download" % self.download_type)
-        download_method(url, tmp_path)
+        # If it is git/wget based download proceed, else fall back if user already
+        # have downloaded tool path, else raise
+        try:
+            download_method = getattr(self, "_%s_download" % self.download_type)
+            download_method(url, tmp_path)
+        except AttributeError:
+            if not self.downloaded_file_path:
+                raise exceptions.TestError("Tool is not downloaded or download"
+                                           " link for Tool is not provided")
+            file_type = process.getoutput("file %s" % self.downloaded_file_path)
```

Comment:
@aliang123 Hi, Thank you for reviewing,

1. my thinking is, when both `downloaded_file_path` and `download_url` are given, download from url should have high priority and override (The user might give both because already downloaded tool might be older version and wants to try downloading latest). If `download_url` is not given, it falls back and check if user given already `downloaded_file_path` and similar in the other way.

2. `downloaded_file_path` is new option to user, if user have developed a new tool which is not hosted or the tool is already downloaded, using this option user can make the framework to deploy it. If the `download_url` is configured as `None` and `downloaded_file_path` is configured along with other stress params specific to the new/downloaded tool the framework adopts it.

Please let me know your thoughts.

Reply:
So @balamuruhans , another question: if the same stress version in url and downloaded_file_path, we still need to download from url?

And if we use downloaded_file_path, but there is no downloaded stress tools, what can we do? Don't we need to check the tools' existance in  downloaded_file_path first, then we can uncompress it?

Reply:
@aliang123, I have checked with help of your suggestion and API looks to be stable in those scenarios, please feel free to correct me.

> And if we use `downloaded_file_path`, but there is no downloaded stress tools, what can we do? 

If we check inside `downloaded_file_path` for stress tools we cannot be sure that the existing files are proper or valid or corrupted, so if there is empty folder and no downloaded stress tool we raise `TestError` as make cmds will fail below if tool is not proper or not available,

https://github.com/avocado-framework/avocado-vt/blob/fa0abe7396c7af147c866ee90af1e7b62a129419/virttest/utils_test/__init__.py#L2015-L2023

> Don't we need to check the tools' existance in downloaded_file_path first, then we can uncompress it?

Okay, the framework will check and handles in better way, if the `downloaded_file_path` is not archive or empty Archive exception is raised in,

https://github.com/avocado-framework/avocado/blob/7f4c6bcf47063567479ee7d135d2da95b32717f3/avocado/utils/archive.py#L119-L123

Reply:
@balamuruhans , i see, thanks for your patient explanation.